{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# AMIA 2016 Annual Symposium Workshop (WG13)\n",
    "\n",
    "## Automated and Scalable Cloud-based RNA-Seq Data Analysis, Part II\n",
    "\n",
    "\n",
    "Riyue Bao, Ph.D. \n",
    "Center for Research Informatics,\n",
    "The University of Chicago.\n",
    "November 13, 2016\n",
    "\n",
    "***\n",
    "\n",
    "## Objective\n",
    "\n",
    "* Explore the quality metrics and input/output of the RNAseq pipeline\n",
    "* Visualize result files and quality plots\n",
    "* Pracice differential gene identification and pathway analysis\n",
    "\n",
    "***\n",
    "\n",
    "## Dataset\n",
    "\n",
    "The test datasets used in this workshop are from \n",
    "Fog. et al. 2015. Loss of PRDM11 promotes MYC-driven lymphomagenesis. Blood, 125(8):1272-81\n",
    "<http://www.bloodjournal.org/content/125/8/1272.long?sso-checked=true>\n",
    "\n",
    "***\n",
    "\n",
    "## Identify DEGs: DESeq2\n",
    "\n",
    "Commonly used tool / package for detection of signficant DEGs include `DESeq2`, `limma/limmavoom`, `edgeR`, etc.\n",
    "\n",
    "For this workshop, We will demo how to use DESeq2 to detect DEGs between WT and KO groups.\n",
    "\n",
    "***\n",
    "\n",
    "### Clean the environment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rm(list=ls())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load libraries / packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- List packages required in this analysis\n",
    "cpan.pkg.list = c('ggplot2', 'scales', 'ape', 'RColorBrewer', 'reshape','VennDiagram')\n",
    "bioc.pkg.list = c('ctc',  'limma', 'edgeR', 'DESeq2', 'vsn', 'genefilter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Set up CPAN repo (required if running IRkernel in Jupyter)\n",
    "cpan.repos = 'http://cran.us.r-project.org'\n",
    "\n",
    "##-- Install CPAN packages\n",
    "# install.packages('ggplot2', repos=cpan.repos) ## included in R essentials\n",
    "# install.packages('scales', repos=cpan.repos) ## included in R essentials\n",
    "# install.packages('ape', repos=cpan.repos)\n",
    "# install.packages('RColorBrewer', repos=cpan.repos)\n",
    "# install.packages('reshape', repos=cpan.repos)\n",
    "# install.packages('VennDiagram', repos=cpan.repos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Bioconductor version 3.3 (BiocInstaller 1.22.3), ?biocLite for help\n"
     ]
    }
   ],
   "source": [
    "##-- Set up Bioconductor repo\n",
    "source(\"http://bioconductor.org/biocLite.R\")\n",
    "\n",
    "##-- Install Bioc packages\n",
    "# biocLite('ctc')\n",
    "# biocLite('limma')\n",
    "# biocLite('edgeR')\n",
    "# biocLite('DESeq2')\n",
    "# biocLite('vsn')\n",
    "# biocLite('genefilter')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Load libraries\n",
    "# lapply(c(cpan.pkg.list, bioc.pkg.list),library, character.only = TRUE)\n",
    "suppressMessages(library('ggplot2'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set up global parameters for the analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rep.flag = 1\n",
    "limma.flag = 1\n",
    "deseq.flag = 0\n",
    "deseq2.flag = 1\n",
    "edger.flag = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Complete workflow\n",
    "```{r}\n",
    "\n",
    "## setting\n",
    "cancer = 'DLBC' \n",
    "fdrs = c(0.01, 0.05)\n",
    "fcs = c(1.5, 2.0) ## 1.4\n",
    "limmatypes = c('limmavoom', 'limmavoomweighted') ## raw counts (not limma itself!)\n",
    "colors=c('cold'='cornflowerblue', 'hot'='lightcoral') ## set tumor group colors\n",
    "sample_ex = c()\n",
    "# sample_ex = c('S1WT06','S4KO05')\n",
    "outputpath = 'diff_expr'\n",
    "\n",
    "outstring = ''\n",
    "if(length(sample_ex) > 0) {\n",
    "  outstring = paste0(sample_ex, collapse = '_')\n",
    "  outstring = paste0('ex',outstring, '.')\n",
    "}\n",
    "\n",
    "## paths\n",
    "path='C:/Users/rbaoc/Desktop/current_proj/CRI_workhop2016/Blood_DLBC/results/degs'\n",
    "setwd(path)\n",
    "\n",
    "## files\n",
    "matrixfile = paste0(cancer, '.raw_counts.txt')\n",
    "groupfile = paste0(cancer, '.sample_group.txt')\n",
    "geneinfofile = 'gencode.v24.primary_assembly.annotation.gtf.geneinfo'\n",
    "\n",
    "indir = '.'\n",
    "outputpath = paste0(indir, '/', outputpath)\n",
    "\n",
    "## print info\n",
    "print(paste0('Cancer = ', cancer))\n",
    "print(paste0('Marix file = ', matrixfile))\n",
    "print(paste0('Sample group file  = ', groupfile))\n",
    "print(paste0('DEG fdr = ', fdrs))\n",
    "print(paste0('DEG fc = ', fcs))\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "## import data \n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "print('Import data ...')\n",
    "\n",
    "## Read data files \n",
    "## note that this expression matrix is already norm + log2 transformed\n",
    "data = read.table(paste0(indir, '/', matrixfile), sep='\\t', header=T, stringsAsFactors=F, na='NA')\n",
    "group = read.table(paste0(indir, '/', groupfile), sep='\\t', header=T, stringsAsFactors=F, na='NA')\n",
    "geneinfo = read.delim(geneinfofile, header = T, stringsAsFactors = F)\n",
    "\n",
    "## remove extra columns \n",
    "data = data[,c(1,7:12)]\n",
    "data[1:3,]\n",
    "\n",
    "colnames(data)[1] = 'Gene'\n",
    "\n",
    "## add gene symbol (for ensemble geneid)\n",
    "print(sum(data$Gene %in% geneinfo$gene_id) == nrow(data))\n",
    "\n",
    "row.names(data) = data[,1]\n",
    "row.names(group) = group[,1]\n",
    "row.names(geneinfo) = geneinfo[,1]\n",
    "\n",
    "data = merge(geneinfo[,c('gene_id', 'gene_name')], data, by = 'row.names')\n",
    "data[1:4,1:5]\n",
    "\n",
    "data$Gene = paste0(data$gene_name, '!', data$gene_id)\n",
    "data = data[,-c(1:3)]\n",
    "data[1:4,1:5]\n",
    "\n",
    "## parse data\n",
    "# colnames(group) = c('Sample', 'Group')\n",
    "row.names(data) = data[,1]\n",
    "data = data[,-1]\n",
    "\n",
    "## remove samples \n",
    "data = data[,! colnames(data) %in% sample_ex]\n",
    "colnames(data)\n",
    "\n",
    "row.names(group) = group$Sample\n",
    "group = group[! group$Sample %in% sample_ex,]\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "## Split by gene type \n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "## split into coding and noncoding \n",
    "\n",
    "head(geneinfo)\n",
    "data.frame(table(geneinfo$gene_type))\n",
    "\n",
    "## retrieve gene id only\n",
    "geneinfo_gene = geneinfo[,c(1,2,4)]\n",
    "geneinfo_gene = unique(geneinfo_gene)\n",
    "dim(geneinfo_gene)\n",
    "\n",
    "geneinfo_gene = data.frame(Key = paste0(geneinfo_gene$gene_name,'!', geneinfo_gene$gene_id), geneinfo_gene)\n",
    "row.names(geneinfo_gene) = geneinfo_gene[,1]\n",
    "geneinfo = geneinfo_gene[,-1]\n",
    "\n",
    "gene_coding = geneinfo_gene[geneinfo_gene$gene_type == 'protein_coding',]\n",
    "gene_lincrna = geneinfo[geneinfo_gene$gene_type == 'lincRNA',]\n",
    "\n",
    "dim(gene_coding)\n",
    "dim(gene_lincrna)\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "## Select cancer\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "group_sub = group\n",
    "\n",
    "type = 'coding'\n",
    "if(type == 'hg19') {data_sub = data  } else {\n",
    "  if(type == 'coding') { gene_sub = gene_coding  }\n",
    "  if(type == 'lincrna') { gene_sub = gene_lincrna }\n",
    "  \n",
    "  data_sub = data[row.names(data) %in% row.names(gene_sub),] \n",
    "}\n",
    "\n",
    "dim(data_sub)\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "## detect DEGs\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "print('Detect DEGs ...')\n",
    "\n",
    "## sort matrix sample col to be consistent with group list\n",
    "data_sub = data_sub[,group_sub$Sample] \n",
    "# data_sub_print = data.frame(Gene = row.names(data_sub), data_sub)\n",
    "# save(data_sub, file=paste0(cancer,'.',matrixfile,'.data.RData'))\n",
    "# write.table(data_sub_print, file=paste0(cancer,'.',matrixfile,'.data.txt'), row.names = F, col.names = T, quote = F, sep = '\\t')\n",
    "\n",
    "## filter \n",
    "dim(data_sub)\n",
    "# keep = rowSums(data_sub>0) >= ((dim(data_sub)[2]) * 0.5) ## at least 50% samples have >0 counts\n",
    "# data_sub = data_sub[keep,] \n",
    "dim(data_sub)\n",
    "## 17313   354\n",
    "# save(data_sub, file=paste0(cancer,'.',matrixfile,'.data.flt.RData'))\n",
    "# data_sub_print = data.frame(Gene = row.names(data_sub), data_sub)\n",
    "# write.table(data_sub_print, file=paste0(cancer,'.',matrixfile,'.data.flt.txt'), row.names = F, col.names = T, quote = F, sep = '\\t')\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "if(limma.flag == 1) {\n",
    "  \n",
    "  print(\"Running limma ... \")\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## filter \n",
    "  dim(data_sub)\n",
    "  keep = rowSums(data_sub>0) >= ((dim(data_sub)[2]) * 0.5) ## at least 50% samples have >0 counts\n",
    "  data_sub_flt = data_sub[keep,] \n",
    "  dim(data_sub_flt)\n",
    "  ## 17313   354\n",
    "  # save(data_sub_flt, file=paste0(cancer,'.',matrixfile,'.data.flt.RData'))\n",
    "  # data_sub_flt_print = data.frame(Gene = row.names(data_sub_flt), data_sub_flt)\n",
    "  # write.table(data_sub_flt_print, file=paste0(cancer,'.',matrixfile,'.data.flt.txt'), row.names = F, col.names = T, quote = F, sep = '\\t')\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## prepare design matrix and contrasts\n",
    "  Group = as.factor(group_sub$Group)\n",
    "  design = model.matrix(~ 0+Group)\n",
    "  colnames(design) = gsub('Dataset', '', gsub('Group','',colnames(design)))\n",
    "  head(design)\n",
    "  \n",
    "  contrast_matrix = makeContrasts(KO - WT,levels=design)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  for(limmatype in limmatypes) {\n",
    "    \n",
    "    print(paste0('limmatype = ', limmatype))\n",
    "    \n",
    "    if( limmatype == 'limma' ) { \n",
    "      ## create an ExpressionSet obj from a matrix (use Biobase)\n",
    "      eset = new('ExpressionSet', exprs=as.matrix(data_sub_flt))\n",
    "      fit = lmFit(eset, design) \n",
    "      \n",
    "      ## ------------------------------------------------------------------------\n",
    "      \n",
    "    } else if ( limmatype == 'limmavoom' | limmatype == 'limmavoomweighted') {\n",
    "      \n",
    "      dge = DGEList(counts=as.matrix(data_sub_flt))\n",
    "      #   dim(dge)\n",
    "      #   keep = rowSums(cpm(dge)>1) >= ((dim(dge)[2]-1) * 0.5) ## at least 50% samples have CPM>=1\n",
    "      #   dge = dge[keep,] \n",
    "      #   dim(dge)\n",
    "      dge = calcNormFactors(dge, method='TMM') \n",
    "      pdf(file=paste0(outputpath, '/', limmatype, '/', cancer,'.',type, '.', outstring,limmatype, '.meanvar.pdf'), width=8, height=6)\n",
    "      if( limmatype == 'limmavoom') {\n",
    "        data_voom = voom(dge, design, normalize.method='none', plot=TRUE)\n",
    "      } else if( limmatype == 'limmavoomweighted') {\n",
    "        data_voom = voomWithQualityWeights(dge, design, normalize.method='none', plot=TRUE)\n",
    "      }\n",
    "      dev.off()\n",
    "      save(data_voom, file=paste0(outputpath, '/',limmatype, '/',cancer,'.', type, '.',outstring, limmatype,'.RData'))\n",
    "      data_voom_matrix = data.frame(Gene = row.names(data_voom$E), round(data_voom$E,2)) ## voom norm matrix\n",
    "      write.table(data_voom_matrix, file=paste0(outputpath, '/',limmatype, '/',cancer,'.', \n",
    "                                                type, '.',outstring,\n",
    "                                                limmatype,'.txt'), \n",
    "                  row.names = F, col.names = T, quote = F, sep = '\\t')\n",
    "      \n",
    "      fit = lmFit(data_voom, design)\n",
    "    }\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    fit2 = contrasts.fit(fit, contrast_matrix)\n",
    "    fit2 = eBayes(fit2)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    coef_colnames = gsub(' - ', 'vs', colnames(fit2$cov.coefficients))\n",
    "    deg_lists = list()\n",
    "    \n",
    "    for (i in 1:length(coef_colnames)) {\n",
    "      print(paste0('cov.coefficients col number = ', i))\n",
    "      comp = coef_colnames[i]\n",
    "      output = paste0(outputpath,'/',limmatype, '/',cancer, '.', type, '.',\n",
    "                      outstring,comp, '.', limmatype, '.txt')\n",
    "      table = topTable(fit2, coef=i, adjust='fdr', number=100000, sort.by='p', resort.by='p', p.value=1, lfc=log2(1.0))\n",
    "      \n",
    "      ## ------------------------------------------------------------------------\n",
    "      \n",
    "      ## add gene name    \n",
    "      table$Gene = row.names(table)\n",
    "      table = table[,c('Gene', colnames(table)[1:6])]\n",
    "      \n",
    "      ## ------------------------------------------------------------------------\n",
    "      \n",
    "      ## anti-log FC\n",
    "      table$FC = 0 \n",
    "      row_pos = which(table$logFC >= 0)\n",
    "      row_neg = which(table$logFC < 0)\n",
    "      table$FC[row_pos] = 2^table$logFC[row_pos]\n",
    "      table$FC[row_neg] = -2^((-1) * table$logFC[row_neg])\n",
    "      print(sum(table$FC == 0) == 0)\n",
    "      head(table)\n",
    "      write.table(table, output, col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "      \n",
    "      ## ------------------------------------------------------------------------\n",
    "      \n",
    "      ## flt DEGs \n",
    "      for(fdr in fdrs) {\n",
    "        for(fc in fcs) {\n",
    "          print(paste0(fc, ',', fdr))\n",
    "          table_flt = table[table$adj.P.Val < fdr & (table$logFC >= log2(fc) | \n",
    "                                                       table$logFC <= -log2(fc)),]\n",
    "          \n",
    "          if(nrow(table_flt) > 0) { \n",
    "            write.table(table_flt, paste0(output, '.flt.fdr', fdr, '_fc', fc), \n",
    "                        col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "            \n",
    "            ## print gene symbol only as input for IPA \n",
    "            genes = data.frame(Gene = gsub('!\\\\S+', '', table_flt$Gene), \n",
    "                               table_flt[,colnames(table_flt) %in% \n",
    "                                           c('logFC', 'P.Value', 'adj.P.Val', 'FC')])\n",
    "            print(sum(genes$FC == 0) == 0)\n",
    "            head(genes)\n",
    "            genes = genes[genes$Gene != '---' & ! is.na(genes$Gene),]\n",
    "            \n",
    "            write.table(genes, paste0(output, '.flt.fdr', fdr, '_fc', fc, '.input'), \n",
    "                        col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "            write.table(genes[genes$logFC > 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                         '.up.input'), \n",
    "                        col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "            write.table(genes[genes$logFC < 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                         '.down.input'), \n",
    "                        col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "            \n",
    "            ## all three statements generate the same output\n",
    "            # deg_lists = append(deg_lists, list(table_flt[,1]))\n",
    "            # deg_lists = c(deg_lists, list(table_flt[,1]))\n",
    "            deg_lists[[i]] = as.vector(table_flt[,1])\n",
    "            \n",
    "          }\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "if(deseq.flag == 1)\n",
    "{\n",
    "  print(\"Running DESeq ... \")\n",
    "  \n",
    "  caller = 'deseq'\n",
    "  \n",
    "  output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                  type, '.',outstring, caller)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "\n",
    "  # detach(\"package:DESeq2\", unload=TRUE)\n",
    "  library(DESeq) ## conflict with DESeq2\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## note group_sub must have the same sample order as data_sub\n",
    "  Group = as.factor(group_sub$Group)\n",
    "  \n",
    "  ## Build a count data set\n",
    "  cds = newCountDataSet( data_sub, Group )\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## estimate lib size factors\n",
    "  cds = estimateSizeFactors( cds )\n",
    "  sizeFactors(cds)\n",
    "  \n",
    "  ## print normalized matrix\n",
    "  cds.before.norm = counts( cds, normalized=F) ## same as data_sub\n",
    "  cds.after.norm = counts( cds, normalized=T )\n",
    "  \n",
    "  cds.before.norm[1:3,1:3]\n",
    "  cds.after.norm[1:3,1:3]\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## log2 transformation with Variance stabilizing transformation to estimate n0\n",
    "  ## NOT for DEG analysis, but for heatmap, clustering etc.\n",
    "  ## it is important to use 'blind' so the dispersion estimation is not skewed towards design\n",
    "  \n",
    "  cdsBlind = estimateDispersions( cds, method=\"blind\" )\n",
    "  vsd = varianceStabilizingTransformation( cdsBlind )\n",
    "  vsd.expr = exprs(vsd)\n",
    "  \n",
    "  save(vsd, file=paste0(output,'.RData'))\n",
    "  \n",
    "  ## save normalized and log2-transformed matrix\n",
    "  write.table(data.frame(Gene = row.names(vsd.expr), vsd.expr),\n",
    "              file = paste0(output,'.txt'),\n",
    "              sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "  \n",
    "  ## plot to confirm that vsn removes mean-variance trend \n",
    "  # par(mfrow=c(1,2))\n",
    "  notAllZero = (rowSums(counts(cds))>0)\n",
    "  \n",
    "  pdf(paste0(output, '.meanvar.log2.pdf'),width = 7, height = 7)\n",
    "  meanSdPlot(log2(counts(cds)[notAllZero, ] + 1))\n",
    "  dev.off()\n",
    "  \n",
    "  pdf(paste0(output, '.meanvar.vsd.pdf'),width = 7, height = 7)\n",
    "  meanSdPlot(vsd[notAllZero, ])\n",
    "  dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## estimate the dispersions for DEG analysis\n",
    "  if(rep.flag == 1) { cds = estimateDispersions(cds) }\n",
    "  if(rep.flag == 0) { cds = estimateDispersions(cds, method=\"blind\", sharingMode=\"fit-only\")}\n",
    "  \n",
    "  head( fData(cds) )\n",
    "  \n",
    "  save(cds, file=paste0(output,'.dispersion.RData'))\n",
    "  \n",
    "  #Plot of estimated dispersions\n",
    "  pdf(paste0(output, '.dispersion.pdf'),width = 7, height = 7)\n",
    "  plotDispEsts(cds)\n",
    "  dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  group1 = 'KO'\n",
    "  group2 = 'WT'\n",
    "  \n",
    "  comp = paste0(group1, 'vs', group2, '.')\n",
    "  output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                  type, '.',outstring, comp,caller,'.txt')\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## filter out low express genes\n",
    "  rs = rowSums ( counts ( cds ))\n",
    "  theta = 0.1 ## fraction fo genes to filter out  \n",
    "  row.select = (rs > quantile(rs, probs=theta))\n",
    "  table(row.select)\n",
    "  ## FALSE  TRUE \n",
    "  ## 5486 16562 \n",
    "  \n",
    "  cds.flt = cds[ row.select, ]\n",
    "  \n",
    "  dim(cds)\n",
    "  dim(cds.flt)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## investigate if the threshold (theta) affects DEG result)\n",
    "  res = nbinomTest(cds, group1, group2)\n",
    "  \n",
    "  res.plot = data.frame(\n",
    "    filterstat = rs,\n",
    "    pvalue = res$pval,\n",
    "    row.names = featureNames(cds) )\n",
    "  head(res.plot)\n",
    "  \n",
    "  theta.test = seq(from=0, to=0.5, by=0.1)\n",
    "  pBH = filtered_p(filter=res.plot$filterstat, \n",
    "                   test=res.plot$pvalue, \n",
    "                   theta=theta.test, \n",
    "                   method=\"BH\")\n",
    "  head(pBH)\n",
    "  \n",
    "  pdf(paste0(output, '.theta.rejection_plot1.pdf'),width = 7, height = 7)\n",
    "  \n",
    "  rejection_plot(pBH, at=\"sample\",\n",
    "                 xlim=c(0, 0.5), ylim=c(0, 2000),\n",
    "                 xlab=\"FDR cutoff (Benjamini & Hochberg adjusted p-value)\", main=\"\")\n",
    "  \n",
    "  dev.off()\n",
    "  \n",
    "  pdf(paste0(output, '.theta.rejection_plot2.pdf'),width = 7, height = 7)\n",
    "  \n",
    "  rejBH = filtered_R(alpha=0.1, \n",
    "                     filter=res.plot$filterstat, \n",
    "                     test=res.plot$pvalue, \n",
    "                     theta=theta.test, \n",
    "                     method=\"BH\")\n",
    "\n",
    "  plot(theta.test, rejBH, type=\"l\",\n",
    "       xlab=expression(theta.test), ylab=\"number of rejections\")\n",
    "  abline(v = theta, col = \"red\", lwd = 3)\n",
    "  \n",
    "  dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## DEG detection (BH for multitesting correction; cannot change ... )\n",
    "  res = nbinomTest(cds.flt, group1, group2);\n",
    "  \n",
    "  save(res, file = paste0(output,'.RData'))\n",
    "  \n",
    "  write.table(res,\n",
    "              file = output,\n",
    "              sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## filter DEGs\n",
    "  \n",
    "  #remove nan values for the foldchange == NAN\n",
    "  dim(res)\n",
    "  res = res[!is.nan(res$foldChange) & ! is.nan(res$padj),];\n",
    "  dim(res)\n",
    "  \n",
    "  for(fdr in fdrs) {\n",
    "    for(fc in fcs) {\n",
    "      print(paste0(fc, ',', fdr))\n",
    "      \n",
    "      #Plot of MA to display differential expression versus expression strength\n",
    "      pdf(paste0(output,'.MAplot.pdf'), width=7, height=7)\n",
    "      #    plotMA(res[!is.infinite(res$log2FoldChange),c(2,6,8)]);\n",
    "      plotMA(res)\n",
    "      dev.off()\n",
    "      \n",
    "      #Plot of histogram of unadj p-value for sanity check\n",
    "      pdf(paste0(output,'.pvalue.hist.pdf'), width=7, height=7)\n",
    "      hist(res$pval, breaks=100, col=\"skyblue\", border=\"slateblue\", main=\"Histogram of unadjusted p-value\", xlab=\"unajusted p-value\") \n",
    "      dev.off()   \n",
    "      \n",
    "      res.flt = res[(res$foldChange >= fc | res$foldChange <= -fc) & \n",
    "                      res$padj < fdr,]\n",
    "      dim(res.flt)\n",
    "      \n",
    "      if(nrow(res.flt) > 0) {\n",
    "        write.table(res.flt, paste0(output, '.flt.fdr', fdr, '_fc', fc), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        \n",
    "        ## print gene symbol only as input for IPA \n",
    "        genes = data.frame(id = gsub('!\\\\S+', '', res.flt$id), \n",
    "                           res.flt[,colnames(res.flt) %in% \n",
    "                                       c('log2FoldChange', 'pval', 'padj', 'foldChange')])\n",
    "        print(sum(genes$foldChange == 0) == 0)\n",
    "        head(genes)\n",
    "        genes = genes[genes$id != '---' & ! is.na(genes$id),]\n",
    "        \n",
    "        write.table(genes, paste0(output, '.flt.fdr', fdr, '_fc', fc, '.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        write.table(genes[genes$log2FoldChange > 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                     '.up.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        write.table(genes[genes$log2FoldChange < 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                     '.down.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "if(deseq2.flag == 1)\n",
    "{\n",
    "  \n",
    "  print(\"Running DESeq2 ... \")\n",
    "  \n",
    "  caller = 'deseq2'\n",
    "  \n",
    "  output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                  type, '.',outstring, caller)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  if(deseq.flag == 1) { detach(\"package:DESeq\", unload=TRUE) }\n",
    "  library(DESeq2) ## conflict with DESeq\n",
    "\n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## note group_sub must have the same sample order as data_sub\n",
    "  Group = as.factor(group_sub$Group)\n",
    "  \n",
    "  ## Build a count matrix\n",
    "  cds = as.matrix(data_sub)\n",
    "  \n",
    "  dds = DESeqDataSetFromMatrix(countData = cds,\n",
    "                               colData = group_sub, \n",
    "                               design = ~ Group)\n",
    "  \n",
    "  # ## note this is just simple filter to reduce mem; no affect on DEG results\n",
    "  # dim(dds)\n",
    "  # dds = dds[ rowSums(counts(dds)) > 0, ]\n",
    "  # dim(dds)\n",
    "  \n",
    "  # #Plot of estimated dispersions\n",
    "  # pdf(paste0(output, '.dispersion.pdf'),width = 7, height = 7)\n",
    "  # plotDispEsts(dds)\n",
    "  # dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "\n",
    "  ## norm matrix for heatmap clusterin etc. three methods ...\n",
    "\n",
    "  rld = rlog(dds, blind=FALSE)\n",
    "  vsd = varianceStabilizingTransformation(dds, blind=FALSE)\n",
    "  # vsd.fast = vst(dds, blind=FALSE)\n",
    "  \n",
    "  head(assay(rld), 3)\n",
    "  \n",
    "  write.table(data.frame(Gene = row.names(assay(rld)), assay(rld)),\n",
    "              file = paste0(output, '.rld.txt'),\n",
    "              sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "  \n",
    "  write.table(data.frame(Gene = row.names(assay(vsd)), assay(vsd)),\n",
    "              file = paste0(output, '.vsd.txt'),\n",
    "              sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## plot mean to var to confirm variance shrink works \n",
    "  notAllZero = (rowSums(counts(dds))>0)\n",
    "  \n",
    "  pdf(paste0(output, '.meanvar.log2.pdf'),width = 7, height = 7)\n",
    "  meanSdPlot(log2(counts(estimateSizeFactors(dds),normalized=TRUE)[notAllZero,] + 1))\n",
    "  dev.off()\n",
    "  \n",
    "  pdf(paste0(output, '.meanvar.rld.pdf'),width = 7, height = 7)\n",
    "  meanSdPlot(assay(rld[notAllZero,]))\n",
    "  dev.off()\n",
    "  \n",
    "  pdf(paste0(output, '.meanvar.vsd.pdf'),width = 7, height = 7)\n",
    "  meanSdPlot(assay(vsd[notAllZero,]))\n",
    "  dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## plot sample correlation \n",
    "  \n",
    "  sampleDists = dist(t(assay(rld)))\n",
    "  \n",
    "  sampleDistMatrix = as.matrix(sampleDists)\n",
    "  rownames(sampleDistMatrix) = paste(rld$Group, rld$Libtype, sep=\"-\")\n",
    "  colnames(sampleDistMatrix) = rownames(sampleDistMatrix)\n",
    "  \n",
    "  colors = rev(cm.colors(32))[1:16]\n",
    "  \n",
    "  pdf(paste0(output, '.sm_cor.pdf'),width = 7, height = 7)\n",
    "  pheatmap(sampleDistMatrix,\n",
    "           clustering_distance_rows=sampleDists,\n",
    "           clustering_distance_cols=sampleDists,\n",
    "           col=colors)\n",
    "  dev.off()\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  group1 = 'KO'\n",
    "  group2 = 'WT'\n",
    "  \n",
    "  comp = paste0(group1, 'vs', group2, '.')\n",
    "  output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                  type, '.',outstring, comp,caller,'.txt')\n",
    "  \n",
    "  ## DEG analysis\n",
    "  dds = DESeq(dds, test=\"Wald\", betaPrior=T)\n",
    "  res = results(dds, contrast=c(\"Group\",group1,group2), pAdjustMethod =\"fdr\", alpha=fdr)\n",
    "  \n",
    "  save(res, file = paste0(output,'.RData'))\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## Plot of MA to display differential expression versus expression strength\n",
    "  pdf(paste0(output,'.MAplot.pdf'), width=7, height=7)\n",
    "  plotMA(res, main=\"DESeq2\", ylim=c(-2,2))\n",
    "  dev.off()\n",
    "  \n",
    "  ## print more info of the test\n",
    "  mcols(res)$description\n",
    "\n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  head(res)\n",
    "  summary(res)\n",
    "  \n",
    "  res = as.data.frame(res)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## anti-log FC\n",
    "  res$foldChange = NA\n",
    "  row_pos = which(! is.na(res$log2FoldChange) & res$log2FoldChange >= 0)\n",
    "  row_neg = which(! is.na(res$log2FoldChange) & res$log2FoldChange < 0)\n",
    "  res$foldChange[row_pos] = 2^res$log2FoldChange[row_pos]\n",
    "  res$foldChange[row_neg] = -2^((-1) * res$log2FoldChange[row_neg])\n",
    "  print(sum(res$foldChange == 0) == 0)\n",
    "  head(res)\n",
    "  \n",
    "  res = data.frame(id = row.names(res), res)\n",
    "\n",
    "  write.table(res,\n",
    "              file = output,\n",
    "              sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "  \n",
    "  ## ------------------------------------------------------------------------\n",
    "  \n",
    "  ## filter DEGs\n",
    "  \n",
    "  #remove nan values for the foldchange == NAN\n",
    "  dim(res)\n",
    "  res = res[!is.na(res$foldChange) & ! is.na(res$padj),];\n",
    "  dim(res)\n",
    "  \n",
    "  for(fdr in fdrs) {\n",
    "    for(fc in fcs) {\n",
    "      print(paste0(fc, ',', fdr))\n",
    "      \n",
    "      #Plot of histogram of unadj p-value for sanity check\n",
    "      pdf(paste0(output,'.pvalue.hist.pdf'), width=7, height=7)\n",
    "      hist(res$pval, breaks=100, col=\"skyblue\", border=\"slateblue\", main=\"Histogram of unadjusted p-value\", xlab=\"unajusted p-value\") \n",
    "      dev.off()   \n",
    "      \n",
    "      res.flt = res[(res$foldChange >= fc | res$foldChange <= -fc) & \n",
    "                      res$padj < fdr,]\n",
    "      dim(res.flt)\n",
    "      \n",
    "      if(nrow(res.flt) > 0) {\n",
    "        write.table(res.flt, paste0(output, '.flt.fdr', fdr, '_fc', fc), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        \n",
    "        ## print gene symbol only as input for IPA \n",
    "        genes = data.frame(id = gsub('!\\\\S+', '', res.flt$id), \n",
    "                           res.flt[,colnames(res.flt) %in% \n",
    "                                     c('log2FoldChange', 'pval', 'padj', 'foldChange')])\n",
    "        print(sum(genes$foldChange == 0) == 0)\n",
    "        head(genes)\n",
    "        genes = genes[genes$id != '---' & ! is.na(genes$id),]\n",
    "        \n",
    "        write.table(genes, paste0(output, '.flt.fdr', fdr, '_fc', fc, '.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        write.table(genes[genes$log2FoldChange > 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                              '.up.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        write.table(genes[genes$log2FoldChange < 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                              '.down.input'), \n",
    "                    col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "## ------------------------------------------------------------------------\n",
    "\n",
    "if(edger.flag == 1)\n",
    "{\n",
    "  print(\"Running egdeR ... \")\n",
    "  \n",
    "  caller = 'edger'\n",
    "  \n",
    "  output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                  type, '.',outstring, caller)\n",
    "  \n",
    "  if (rep.flag == 1)\n",
    "  {\n",
    " \n",
    "    Group = as.factor(group_sub$Group)\n",
    "    \n",
    "    counts = as.matrix(data_sub)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    #computes counts permilion with cpm function \n",
    "    cpms = cpm(counts, log=F) \n",
    "    \n",
    "    #In edgeR, it is recommended to remove features without at least 1 read per million in n of the samples, where n is the size of the smallest group of replicates\n",
    "    \n",
    "    keep = rowSums(cpms >1) >= min(data.frame(table(Group))[,2])\n",
    "    \n",
    "    dim(counts)\n",
    "    counts = counts[keep,]\n",
    "    dim(counts)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    #create a DGEList object\n",
    "    d = DGEList(counts=counts, group=Group) \n",
    "    \n",
    "    #estimate normalization factors\n",
    "    d = calcNormFactors(d, method='TMM');\n",
    "    \n",
    "    #estimate common negative binomial dispersion by conditional maximum likelihood\n",
    "    ## might need to add one more flag for this argument in estimateCommonDisp rowsum.filter= value for the filtering out low abundance tags in the est of the common dispersion\n",
    "    d = estimateCommonDisp(d, tol=1e-6);\n",
    "    #estimate empirical bayes tagwise dispersion values\n",
    "    d = estimateTagwiseDisp(d, trend=\"loess\", method=\"grid\")\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    ## print norm matrix\n",
    "    logcpm = cpm(d, prior.count=2, log=TRUE)\n",
    "    \n",
    "    save(logcpm, file = paste0(output, '.RData'))\n",
    "    \n",
    "    write.table(data.frame(Gene = row.names(logcpm), logcpm),\n",
    "                file = paste0(output, '.txt'),\n",
    "                sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    #mean-variance plot. visual representation of the mean-variance relationship with plotMeanVar\n",
    "    pdf(paste0(output,'.meanvar.pdf'), width=7, height=7)\n",
    "    plotMeanVar(d, show.tagwise.vars=T, NBline=TRUE);\n",
    "    dev.off()\n",
    "    \n",
    "    #mean-variance plot. visual representation of the mean-variance relationship with plotBCV\n",
    "    pdf(paste0(output,'.bcv.pdf'), width=7, height=7)\n",
    "    plotBCV(d);\n",
    "    dev.off()  \n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    group1 = 'KO'\n",
    "    group2 = 'WT'\n",
    "    \n",
    "    comp = paste0(group1, 'vs', group2, '.')\n",
    "    output = paste0(outputpath, '/', caller,'/',cancer,'.', \n",
    "                    type, '.',outstring, comp,caller,'.txt')\n",
    "    \n",
    "    de = exactTest(d, pair=c(group1, group2));\n",
    "    \n",
    "    #adjust.method = \"BH\",\"BY\",\"fdr\". FWER method = \"holm\", \"hochberg\", \"hommel\", \"bonferroni\"\n",
    "    tt = topTags(de, n=nrow(d), adjust.method =\"fdr\")  \n",
    "    \n",
    "    rn = rownames(tt$table);\n",
    "    \n",
    "    res = tt$table\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    head(res)\n",
    "    summary(res)\n",
    "    \n",
    "    res = as.data.frame(res)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    ## anti-log FC\n",
    "    res$FC = NA\n",
    "    row_pos = which(! is.na(res$logFC) & res$logFC >= 0)\n",
    "    row_neg = which(! is.na(res$logFC) & res$logFC < 0)\n",
    "    res$FC[row_pos] = 2^res$logFC[row_pos]\n",
    "    res$FC[row_neg] = -2^((-1) * res$logFC[row_neg])\n",
    "    print(sum(res$FC == 0) == 0)\n",
    "    head(res)\n",
    "    \n",
    "    res = data.frame(Gene = row.names(res), res)\n",
    "    \n",
    "    write.table(res,\n",
    "                file = output,\n",
    "                sep = '\\t', col.names = T, row.names = F, quote = F)\n",
    "    \n",
    "    ## ------------------------------------------------------------------------\n",
    "    \n",
    "    ## filter DEGs\n",
    "    \n",
    "    #remove nan values for the foldchange == NAN\n",
    "    dim(res)\n",
    "    res = res[!is.na(res$FC) & ! is.na(res$FDR),];\n",
    "    dim(res)\n",
    "    \n",
    "    for(fdr in fdrs) {\n",
    "      for(fc in fcs) {\n",
    "        print(paste0(fc, ',', fdr))\n",
    "        \n",
    "        #Plot of histogram of unadj p-value for sanity check\n",
    "        pdf(paste0(output,'.pvalue.hist.pdf'), width=7, height=7)\n",
    "        hist(res$PValue, breaks=100, col=\"skyblue\", border=\"slateblue\", main=\"Histogram of unadjusted p-value\", xlab=\"unajusted p-value\") \n",
    "        dev.off()   \n",
    "        \n",
    "        #creat a graphical summary for DEG\n",
    "        pdf(file = paste0(output,'.flt.fdr',fdr,'.pdf') ,width=7,height=7)    \n",
    "        deg = rn[tt$table$FDR < fdr];\n",
    "        plotSmear(d, de.tags=deg)\n",
    "        dev.off()\n",
    "        \n",
    "        res.flt = res[(res$FC >= fc | res$FC <= -fc) & \n",
    "                        res$FDR < fdr,]\n",
    "        dim(res.flt)\n",
    "        \n",
    "        if(nrow(res.flt) > 0) {\n",
    "          write.table(res.flt, paste0(output, '.flt.fdr', fdr, '_fc', fc), \n",
    "                      col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "          \n",
    "          ## print gene symbol only as input for IPA \n",
    "          genes = data.frame(id = gsub('!\\\\S+', '', res.flt$Gene), \n",
    "                             res.flt[,colnames(res.flt) %in% \n",
    "                                       c('logFC', 'PValue', 'FDR', 'FC')])\n",
    "          print(sum(genes$FC == 0) == 0)\n",
    "          head(genes)\n",
    "          genes = genes[genes$id != '---' & ! is.na(genes$id),]\n",
    "          \n",
    "          write.table(genes, paste0(output, '.flt.fdr', fdr, '_fc', fc, '.input'), \n",
    "                      col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "          write.table(genes[genes$logFC > 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                                '.up.input'), \n",
    "                      col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "          write.table(genes[genes$logFC < 0, ], paste0(output, '.flt.fdr', fdr, '_fc', fc, \n",
    "                                                                '.down.input'), \n",
    "                      col.names=T, row.names=F, sep='\\t', quote=F)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \n",
    "  }\n",
    "}\n",
    "\n",
    "#########################################################\n",
    "\n",
    "print('...Done!')\n",
    "\n",
    "sessionInfo()\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
